hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageTextWriter.java

```
@@ -651,14 +683,123 @@ private void output(Configuration conf, FileSummary summary,
+    for (int i = 0; i < nThreads; i++) {
+      // Each thread processes different ordered sub-sections
+      // and outputs to different paths
+      int step = subSections.size() / nThreads +
+          (i < subSections.size() % nThreads ? 1 : 0);
+      int start = mark;
+      int end = start + step;
+      ArrayList<FileSummary.Section> subList = new ArrayList<>(
+          subSections.subList(start, end));
+      mark = end;
+      String path = paths[i];
+
+      threads[i] = new Thread(() -> {
+        LOG.info("output iNodes of sub-sections: [{},{})", start, end);
+        InputStream is = null;
+        try (PrintStream theOut = new PrintStream(path, "UTF-8")) {
+          long startTime = Time.monotonicNow();
+          for (int j = 0; j < subList.size(); j++) {
+            is = getInputStreamForSection(subList.get(j), codec, conf);
+            if (start == 0 && j == 0) {
+              // The first iNode section has a header which must be
+              // processed first
+              INodeSection s = INodeSection.parseDelimitedFrom(is);
+              expectedINodes.set(s.getNumInodes());
+            }
+            totalParsed.addAndGet(outputINodes(is, theOut));
+          }
+          long timeTaken = Time.monotonicNow() - startTime;
+          LOG.info("Time to output iNodes of sub-sections: [{},{}) {} ms",
+              start, end, timeTaken);
+        } catch (Exception e) {
+          exceptions.add(new IOException(e));
+        } finally {
+          try {
+            is.close();
+          } catch (IOException ioe) {
+            LOG.warn("Failed to close the input stream, ignoring", ioe);
+          }
+        }
+      });
+    }
+    for (Thread t : threads) {
+      t.start();
+    }
+    for (Thread t : threads) {
+      try {
+        t.join();
+      } catch (InterruptedException e) {
+        LOG.error("Interrupted when thread join.", e);
+        throw new IOException(e);
+      }
+    }
+
+    if (exceptions.size() != 0) {
```

Is this check happening after all threads are finished? Can we return the exception promptly?